test/k8s-integration/main.go

```diff
@@ -157,69 +178,134 @@ func handle() error {
 		}
 	}
 
+	// Create temporary directories for kubernetes builds
+	k8sParentDir := getTempDir()
+	k8sDir := filepath.Join(k8sParentDir, "kubernetes")
+	testParentDir := getTempDir()
+	testDir := filepath.Join(testParentDir, "kubernetes")
+	defer removeTempDir(k8sParentDir)
+	defer removeTempDir(testParentDir)
+
+	numTasks := 4
+	errChan := make(chan error, numTasks)
+	k8sDependencyChan := make(chan bool, 1)
+
+	// If kube version is set, then download and build Kubernetes for cluster creation
+	// Otherwise, either GKE or a prebuild local K8s dir is being used
+	if len(*kubeVersion) != 0 {
+		go func() {
+			err := downloadKubernetesSource(pkgDir, k8sParentDir, *kubeVersion)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to download Kubernetes source: %v", err)
+				k8sDependencyChan <- false
+				return
+			}
+			err = buildKubernetes(k8sDir)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to build Kubernetes: %v", err)
+				k8sDependencyChan <- false
+				return
+			}
+			k8sDependencyChan <- true
+			errChan <- nil
+		}()
+	} else {
+		errChan <- nil
+		k8sDir = *localK8sDir
+		k8sDependencyChan <- true
+	}
+
+	// If test version is set, then download and build Kubernetes to run K8s tests
+	// Otherwise, either kube version is set (which implies GCE) or a local K8s dir is being used
+	if len(*testVersion) != 0 && *testVersion != *kubeVersion {
+		go func() {
+			// TODO: Build only the tests
+			err := downloadKubernetesSource(pkgDir, testParentDir, *testVersion)
```

Comment:
can we download only one version of the source and just copy it if we need it for test dir

Reply:
I think it makes it a lot more complicated dependency wise, so I'll think about it.